runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/artifact/BeamFileSystemArtifactStagingService.java

```diff
@@ -0,0 +1,292 @@
+      }
+      try (WritableByteChannel manifestWritableByteChannel = FileSystems
+          .create(manifestResourceId, MimeTypes.TEXT)) {
+        manifestWritableByteChannel
+            .write(CHARSET.encode(JsonFormat.printer().print(proxyManifestBuilder.build())));
+      }
+      // TODO: Validate integrity of staged files.
+      responseObserver.onNext(CommitManifestResponse.newBuilder()
+          .setStagingToken(manifestResourceId.toString())
+          .build());
+      responseObserver.onCompleted();
+    } catch (IOException e) {
+      // TODO: Cleanup all the artifacts.
+      LOG.error("Unable to commit manifest.", e);
+      responseObserver.onError(e);
+    }
+  }
+
+  @Override
+  public void close() throws Exception {
+    // Nothing to close here.
+  }
+
+  /**
+   * Generate a stagingSessionToken compatible with {@link BeamFileSystemArtifactStagingService}.
+   *
+   * @param sessionId Unique sessionId for artifact staging.
+   * @param basePath Base path to upload artifacts.
+   * @return Encoded stagingSessionToken.
+   */
+  public static String generateStagingSessionToken(String sessionId, String basePath)
+      throws Exception {
+    StagingSessionToken stagingSessionToken = new StagingSessionToken();
+    stagingSessionToken.setSessionId(sessionId);
+    stagingSessionToken.setBasePath(basePath);
+    return encodeStagingSessionToken(stagingSessionToken);
+  }
+
+  private String encodedFileName(ArtifactMetadata artifactMetadata) {
+    return "artifact_" + Hashing.sha256().hashString(artifactMetadata.getName(), CHARSET)
+        .toString();
+  }
+
+  private StagingSessionToken decodeStagingSessionToken(String stagingSessionToken)
+      throws IOException {
+    try {
+      return MAPPER.readValue(stagingSessionToken, StagingSessionToken.class);
+    } catch (IOException e) {
+      try {
+        LOG.error(
+            "Unable to deserialize staging token {}. Expected format {}. Error {}",
+            stagingSessionToken, MAPPER.writeValueAsString(new StagingSessionToken()),
+            e.getMessage());
+      } catch (JsonProcessingException e1) {
+        LOG.error("Error {} occurred while serializing {}.", e.getMessage(),
+            StagingSessionToken.class);
+      }
+      throw e;
+    }
+  }
+
+  private static String encodeStagingSessionToken(StagingSessionToken stagingSessionToken)
+      throws Exception {
+    try {
+      return MAPPER.writeValueAsString(stagingSessionToken);
+    } catch (JsonProcessingException e) {
+      LOG.error("Error {} occurred while serializing {}.", e.getMessage(),
+          StagingSessionToken.class);
+      throw e;
+    }
+  }
+
+  private ResourceId getJobDirResourceId(String stagingSessionToken) throws IOException {
+    ResourceId baseResourceId;
+    StagingSessionToken parsedToken = decodeStagingSessionToken(stagingSessionToken);
+    try {
+      baseResourceId = FileSystems.matchSingleFileSpec(parsedToken.getBasePath())
```

Comment:
Do we need this codepath at all - can we just always use path from line 163?

It also looks like this method is called quite often, so it'd be good to avoid a FileSystems.match() call every time it's called.

Reply:
Makes sense.

I was under the impression that matchNewResource will throw exception for already existing resource.